scripts/build.py

```diff
@@ -174,47 +184,113 @@ def _ensure_files_exist(filepaths):
         filepaths: list(str). Paths to files that we want to ensure exist.
 
     Raises:
-        Exception: Raised if one or more of the files does not exist.
+        OSError: One or more of the files does not exist.
     """
     for filepath in filepaths:
         if not os.path.isfile(filepath):
-            raise Exception('File %s does not exist.' % filepath)
+            raise OSError('File %s does not exist.' % filepath)
+
+
+def get_file_count(directory_path):
+    """Count total number of file directory, subtracting ignored files.
+
+    Args:
+        directory_path: str. Directory to be walked.
 
+    Returns:
+        int. Total number of files minus ignored files.
+    """
+    total_file_count = 0
+    for _, _, files in os.walk(directory_path):
+        total_file_count += len(files)
+        for filename in files:
+            # Ignore files with certain extensions.
+            if any(filename.endswith(p)
+                   for p in FILE_EXTENSIONS_TO_IGNORE):
+                total_file_count -= 1
+    if directory_path == ASSETS_DEV_DIR:
+        # Add 1 to account for the added hashes.js.
+        total_file_count += 1
+    return total_file_count
 
-def process_html(source_path, target_path, file_hashes):
+
+def _compare_file_count(source_path, target_path):
+    """Ensure that two dir's file counts match.
+
+    Args:
+       source_path: str. Source of staging files.
+       target_path: str. Final filepath or built file's directory.
+
+    Raises:
+        ValueError: Raised if both dir do not have the same file count.
+    """
+    source_dir_file_count = get_file_count(source_path)
+    target_dir_file_count = get_file_count(target_path)
+    if source_dir_file_count != target_dir_file_count:
+        print 'Comparing %s vs %s' % (source_path, target_path)
+        raise ValueError(
+            '%s files in source dir != %s files in target dir.' % (
+                source_dir_file_count, target_dir_file_count))
+
+
+def _match_filename_with_hashes(filename, file_hashes):
+    """Ensure that filepath is hashed correctly.
+
+    Args:
+        filename: str. Filepath to be matched.
+        file_hashes: dict(str, str). Dictionary of file hashes.
+
+    Raises:
+        ValueError: Raised if hash dict is empty.
+        ValueError: Raised if filename does not contain hash.
+        KeyError: Raised if filename's hash does not match hash dict entries.
+    """
+    # Final filepath example: base.240933e7564bd72a4dde42ee23260c5f.html.
+    if not file_hashes:
+        raise ValueError('Hash dict is empty')
+    file_hash = re.findall(r"([a-fA-F\d]{32})", filename)
```

Comment:
Note: we use single quotes in python.

Also this doesn't seem foolproof. Can you use a more deterministic way to get the file hash? E.g. reading the string that comes between the last and second-to-last "." character?

Reply:
Previously, I did attempted to use the the `.` as the delimiters for my regex.
However, there are a few edge cases of filenames such as:

`embedding_tests_jsdelivr_0.0.2.min.2eb183c2355cb1591284c206e4c55051.html`

Thus I resorted to use the above regex, which is a regex only for finding MD5 hashes.
If a different hashing technique is used, it would then fall through. Just curious, what makes it doesn't seem foolproof to, @seanlip ?

Reply:
Not sure what's wrong with that filename? You can still get the part of the string that's between the last and second-to-last dot.

You don't even need to use regex, just do s.split('.') and take s[-2]. Though you'll want to double-check that s[-2] actually exists (and perhaps that it is a 32-char string with chars in [a-fA-f\d]).

The original doesn't seem foolproof to names of files that happen to have 32 char substrings of base64 chars (even without the added hashes).